planning/behavior_path_planner/autoware_behavior_path_lane_change_module/src/utils/utils.cpp

```diff
@@ -1247,4 +1252,57 @@ bool filter_target_lane_objects(
 
   return false;
 }
+
+std::vector<lanelet::ConstLanelets> get_preceding_lanes(const CommonDataPtr & common_data_ptr)
+{
+  const auto & route_handler_ptr = common_data_ptr->route_handler_ptr;
+  const auto & target_lanes = common_data_ptr->lanes_ptr->target;
+  const auto & ego_pose = common_data_ptr->get_ego_pose();
+  const auto backward_lane_length = common_data_ptr->lc_param_ptr->backward_lane_length;
+
+  const auto preceding_lanes_list =
+    utils::getPrecedingLanelets(*route_handler_ptr, target_lanes, ego_pose, backward_lane_length);
+
+  const auto & current_lanes = common_data_ptr->lanes_ptr->current;
+  std::unordered_set<lanelet::Id> current_lanes_id;
+  for (const auto & lane : current_lanes) {
+    current_lanes_id.insert(lane.id());
+
+    const auto prev_lanes = route_handler_ptr->getPreviousLanelets(lane);
+    for (const auto & prev_lane : prev_lanes) {
+      current_lanes_id.insert(prev_lane.id());
+    }
+  }
+
+  const auto trim_lanes = [&](const lanelet::ConstLanelets & preceding) {
+    // Step 1: Remove lanes with the same ID as in `current_lanes_id`
+    auto non_overlapping_lanes =
+      preceding | ranges::views::remove_if([&](const lanelet::ConstLanelet & lane) {
+        return current_lanes_id.find(lane.id()) != current_lanes_id.end();
+      });
+
+    // Step 2: Identify disconnected lanes
+    std::unordered_set<lanelet::Id> not_connected_ids;
+    for (const auto [current_lane, next_lane] : ranges::views::zip(
+           non_overlapping_lanes, non_overlapping_lanes | ranges::views::drop(1))) {
+      const auto next_lanes = route_handler_ptr->getNextLanelets(current_lane);
+
+      const auto is_connected = ranges::any_of(
+        next_lanes, [next_id = next_lane.id()](const auto & lane) { return lane.id() == next_id; });
+
+      if (!is_connected) {
+        not_connected_ids.insert(current_lane.id());
+      }
+    }
+
+    // Step 3: Remove disconnected lanes
+    auto connected_lanes =
+      non_overlapping_lanes | ranges::views::remove_if([&](const lanelet::ConstLanelet & lane) {
+        return not_connected_ids.find(lane.id()) != not_connected_ids.end();
+      });
```

Comment:
Can you explain whats the purpose of this?

Reply:
When we removed the overlapping lanes, we might ended up lanes that is not connected. For example

Original: L4, L3, L2, L1. Assume overlapping is L2.
Overlapping removed: L1, L3, L4

Since L1 and L3 is not directly connected, we should remove L1 from the list, which yield
Connected: L4, L3.

Reply:
@mkquda 
fixed in https://github.com/autowarefoundation/autoware.universe/pull/9526/commits/744e6aa6820155abd1b6bdb80ee125915789ba18